Implement project management commands

fsouza opened the pull request:
I'd recommend running tsuru locally with [tsuru-bootstrap](https://github.com/tsuru/tsuru-bootstrap) to review this locally.

Here's a terminal session using the commands, to help reviewing:

```
% tsuru-admin pool-add -p dev/dev.example.com
Pool successfully registered.
% tsuru-admin pool-add -p qa/qa.example.com
Pool successfully registered.
% tsuru-admin pool-add -p stage/stage.example.com
Pool successfully registered.
% tsuru-admin pool-add -p production/example.com
Pool successfully registered.
% tranor target-set https://gist.githubusercontent.com/fsouza/ac50e24d18bd7c24588164481fa3686b/raw/482b60f8ecff0cf77bcc2db8903dff5492910447/
Target successfully defined!
% tranor env-list
+-------------+-------------------+
| Environment | DNS Suffix        |
+-------------+-------------------+
| dev         | dev.example.com   |
| qa          | qa.example.com    |
| stage       | stage.example.com |
| production  | example.com       |
+-------------+-------------------+
% tranor project-create --name myblog --platform python --team admin
successfully created the project "myblog"!
Git repository: git@192.168.50.4.nip.io:myblog-dev.git
% tranor project-create -n website -l static -t admin -e dev,production
successfully created the project "website"!
Git repository: git@192.168.50.4.nip.io:website-dev.git
% tranor project-list
+---------+--------------+--------------------------+
| Project | Environments | Address                  |
+---------+--------------+--------------------------+
| myblog  | dev          | myblog.dev.example.com   |
|         | qa           | myblog.qa.example.com    |
|         | stage        | myblog.stage.example.com |
|         | production   | myblog.example.com       |
+---------+--------------+--------------------------+
| website | dev          | website.dev.example.com  |
|         | production   | website.example.com      |
+---------+--------------+--------------------------+
# ... manually deploy the app website-dev
% tranor project-info -n website
Project name: website

+-------------+-------------------------+---------+-------------------------------+
| Environment | Address                 | Version | Deploy date                   |
+-------------+-------------------------+---------+-------------------------------+
| dev         | website.dev.example.com | v1      | Mon, 05 Sep 2016 16:36:19 UTC |
| production  | website.example.com     |         |                               |
+-------------+-------------------------+---------+-------------------------------+
```

Commit:
Add platform-list

The user needs to list platforms before being able to create a project.

Commit:
main: register team related commands

Teams are also required to create projects, so we're going to need them.

Commit:
env: add config method to extract name of environments

Commit:
main: register plan-list

Plan is also required for creating applications.

Commit:
env: add method to get pool name from environment

Commit:
project: add command for creating new projects

Commit:
project: add validation in project-create

Commit:
project: add command project-remove

Commit:
project: add the project-list command

Commit:
project: add output to project-remove

Commit:
Change project formatting

Commit:
project: add command project-info

Commit:
Add some tests to corner cases in the app functions

Just ensuring coverage and protecting from future refactoring, the code
itself already worked as expected.

Commit:
app: fix lastDeploy

That's what happens when you mock everything. My bad :-(

Commit:
Improve server mocking

Looks like this doesn't work like that on Linux.

Commit:
Use the table in project-info test

This ensure the same output in the CI.

Commit:
Work around vet bug in Go 1.5 and 1.6

This is a bug that has been fixed in Go 1.7, but we still support those
old releases.

Commit:
Revert "travis: remove  tip from allowed failures"

I have no idea of what's going on.

This reverts commit 2da3b70f3f4e5e54ee995df3860def246a372f12.

aflag commented on project_test.go:
```diff
@@ -771,7 +771,7 @@ func TestProjectInfo(t *testing.T) {
 		t.Fatal(err)
 	}
-	table := cmd.Table{Headers: cmd.Row{"Environment", "Address", "Version", "Deploy date"}}
+	table := cmd.Table{Headers: cmd.Row([]string{"Environment", "Address", "Version", "Deploy date"})}
```
I'm curious, what is the difference between both statements? And what is the bug exactly?

aflag commented on app.go:
```diff
@@ -0,0 +1,52 @@
+	var app map[string]string
+	err = json.NewDecoder(resp.Body).Decode(&app)
+	return app, err
+}
```
It seems this function does not do anything tsuru wouldn't do during app creation. Does it remove some behavior that tsuru does which shouldn't be done?

tatiana replied:
I believe we may be overlooking something - which @fsouza will explain to us :)
But the same comment applies to `deleteApps` & `listApps`.

fsouza replied:
The tsuru command line doesn't offer any library functions, only the commands themselves, so the commands are coupled to the shell environment. For example, this is the "function" that creates an app: https://github.com/tsuru/tsuru-client/blob/6956485efa992d03b5606ae2dd1c3359f4fde1fb/tsuru/client/apps.go#L99. So, in order to get the repository, we would either parse the output or make the app-info request.

I can revisit this and see how much we can reuse.

tatiana replied:
I believe, ideally, Tsuru would have the concepts we need it to have ;) This way we wouldn't even need Tranor.

Considering that's not a reality, the second best scenario would be if the Tsuru client allowed us to extend things without having to parse the output. So, if I had to vote, I'd suggest we contributed to tsuru client exposing things through "functions".

However, given our priorities, I'm ok keeping the way it is - so we can focus on things that are more important.

tatiana commented:
@fsouza two initial comments:
- Would it be possible for us to have a separate PR with the go dependencies? As we currently use Github for reviewing, it got a bit lost with the amount of files which were added
- The documentation you wrote as part of the PR is great - what do you think of adding it as a doc for developers in the project, as a md?

ehault commented:
congratulation @fsouza for beating the previous record ( @aflag : 121 files )

tatiana commented on app.go:
```diff
@@ -0,0 +1,139 @@
+	url, err := cmd.GetURL("/apps")
+	if err != nil {
+		return nil, err
+	}
```
Out of curiosity: why do we need to retrieve `apps` before creating an `app`?

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	return createdApps, nil
+}
+
+func (c *projectCreate) setCNames(apps []map[string]string, client *cmd.Client) error {
```
Same as above. Isn't there a function on tsuru that does just that?

tatiana replied:
@aflag I suppose @fsouza answered this in another comment - apparently Tsuru client doesn't expose this through "nice" functions, and it would be very ugly for us to call the Tsuru command and parse the output.

aflag commented on app.go:
```diff
@@ -0,0 +1,139 @@
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNoContent {
+		return nil, nil
```
Wouldn't it be better if this was an empty list instead of nil?

fsouza replied:
It doesn't make any difference (except that the empty list requires allocation, but we don't care that much). Using a nil slice is more common Go idiom.

aflag replied:
Ok then. I'm all for keeping the common idioms! :)

aflag commented on app.go:
```diff
@@ -0,0 +1,139 @@
+	defer resp.Body.Close()
+	var deploys []deploy
+	if resp.StatusCode == http.StatusNoContent {
+		return d, nil
```
This d will be nil, right? Shouldn't it be explicit?

fsouza replied:
Nops, it's never nil. A struct cannot be nil.

aflag replied:
Hm, I see. It is an empty struct. Is that also the usual way to represent a not found record in go?

fsouza replied:
Usually the error represents that, but in our case an empty deployment and a deployment that haven't happened are the same (we're just displaying some information from the last deployment). Using a pointer and nil would require an if.

aflag commented on app_test.go:
```diff
@@ -0,0 +1,132 @@
+	if err == nil {
+		t.Error("unexpected <nil> error")
+	}
+}
```
There were no tests in which create app, deploy and list apps actually succeeded doing something.

fsouza replied:
They're tested through project related tests, but I'll add more tests to that.

fsouza replied:
Just added more tests, please take another look.

ehault commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	fs   *gnuflag.FlagSet
+}
+
+func (c *projectRemove) Info() *cmd.Info {
```
Since this is Open source and all shouldn't it be better to have all names and desc in a separate file? for editing/translating purpose?

aflag replied:
If the rest of tsuru is already localized I think we should keep it like that, otherwise, I'm not sure the effort is worth it :(

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	return nil
+}
+
+func (c *projectCreate) filterEnvironments(envs []Environment, names []string) []Environment {
```
Maybe getEnvironmentByName would be better?

tatiana replied:
Considering a slice is returned, I vote for keeping the name `filterEnvironments`.

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	for i, envName := range envNames {
+		appNames[i] = fmt.Sprintf("%s-%s", c.name, envName)
+	}
+	errs, err := deleteApps(appNames, client)
```
So I suppose cnames are deleted automatically when the app is deleted?

fsouza replied:
Yep, the cname belongs to the app.

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	for _, app := range apps {
+		row := cmd.Row{app.Env.Name, app.Addr, "", ""}
+		if deploy, err := lastDeploy(client, app.Name); err == nil && deploy.Image != "" {
+			row[2] = deploy.Image
```
So version is actually the image name, right? Why not call it image?

fsouza replied:
Will rename.

tatiana replied:
But isn't `version` how Tsuru calls it? I believe we should keep as consistent as possible with Tsuru nomenclature - `project` and `envs` were concepts we really needed (use) and Tsuru didn't offer. Despite this, I'd suggest we keep to default Tsuru naming. Therefore :+1:  to keeping `version`.

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	if err != nil {
+		return nil, errors.New("unable to load environments file, please make sure that tranor is properly configured")
+	}
+	apps, err := listApps(client, map[string]string{"name": c.name})
```
What does {"name": c.Name} means? Get all apps which start with that name?

fsouza replied:
Get all apps that have that `c.name` in any part of the `name`.

aflag replied:
Would `"^" + c.Name` work? Would it be better?

fsouza replied:
Yes, will update. Thanks!

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+			}
+			nameParts := env.nameRegexp().FindStringSubmatch(app.Name)
+			dnsParts := env.dnsRegexp().FindStringSubmatch(app.CName[0])
+			if len(nameParts) == 2 && len(dnsParts) == 2 && nameParts[1] == dnsParts[1] && nameParts[1] == c.name {
```
I liked this `if` :) You are making very sure that this is really an tranor-friendly app, right? I think this could be a separate function, though.

tatiana commented on project.go:
```diff
@@ -0,0 +1,392 @@
+		return fmt.Errorf("failed to load environments: %s", err)
+	}
+	envs := c.filterEnvironments(config.Environments, c.envs.Values())
+	apps, err := c.createApps(envs, client)
```
If someone tried to create a project with the same name but in different environments, it seems tranor would allow it, wouldn't it? Eg:

```
tranor project-create --name xubiru --envs dev
tranor project-create --name xubiru --envs qa
```

Perhaps we want to deal with this somehow.

fsouza replied:
Oh, indeed. Should we check that the project doesn't exist before creating the apps?

tatiana replied:
:+1: yes, @fsouza,  I'd say so :)

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+				app.Addr = app.CName[0]
+				app.Env = env
+				projectApps = append(projectApps, app)
+				continue
```
I believe this continue is not necessary.

fsouza replied:
ops, probably some code that I deleted. Thanks!

ehault commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	if err != nil {
+		return nil, err
+	}
+	var projectApps []app
```
This whole part

tatiana replied:
sorry, @Lepthy, what does this mean?

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+			}
+			partsName := env.nameRegexp().FindStringSubmatch(app.Name)
+			partsDNS := env.dnsRegexp().FindStringSubmatch(app.CName[0])
+			if len(partsName) == 2 && len(partsDNS) == 2 && partsName[1] == partsDNS[1] {
```
This could use the function created for the previous comment. I think it could be called `isTranorApp` or something.

fsouza replied:
They're not the same though. That if extracts the project name from the app name and check if the project name

This if only extracts the project name. But I can have function to extract the name (returning `string, error`). It owuld return the project name or an error if the app don't match.

aflag replied:
I think that approach is better :)

fsouza replied:
Cool, will do that. Thanks!

ehault commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	if err != nil {
+		return err
+	}
+	projects := make(map[string][]app)
```
looks awfully similar to this one, except the reversed variable name (dnsPart vs partns), couldn't that be factored?

fsouza replied:
We'd have to create a helper function that takes the client and the app filter and returns `[]app, config, err` (something like `<fnname>(client *cmd.Client, filters map[string]string) ([]app, config, err)`). Naming the function is also hard (`getConfigAndApps`?), because it doesn't have a clear reason to exist (except refactoring).

I think it would be worse to do that, but I'm happy to refactor if everyone agrees. Thoughts?

tatiana replied:
+1 to keep the way it is now

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	sort.Sort(list)
+	var table cmd.Table
+	table.LineSeparator = true
+	table.Headers = cmd.Row{"Project", "Environments", "Address"}
```
Could we also display the plan for each environment?

fsouza replied:
Tsuru doesn't deliver this in app-list, unfortunately. Only in app-info.

aflag commented on project.go:
```diff
@@ -0,0 +1,392 @@
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(ctx.Stdout, "Project name: %s\n\n", c.name)
```
I'm missing the platform, plan, teams, etc. Was the idea to use tsuru app-info directly for those things? What about the units?

fsouza replied:
Oh, nice catch. We should include everything here. I'll update the code.

tatiana replied:
I suggest `project-list` contains additional platform + teams.

Just talked to Rafa and we propose this output for `project-info`

```
% tranor project-info -n website

Project name: website
Description:
Repository: git@git.tsuru.example.com:website.git
Platform: ruby
Teams: example-team
Owner: someone@example.com
Team owner: example-team


+-------------+-------------------------+-------+---------------+-------------------------------+-------+
| Environment | Address                 | Image | Git hash /tag |Deploy date                    | Units |
+-------------+-------------------------+-------+---------------+----------------+--------------+-------+
| dev         | website.dev.example.com | v2    | (git) 7d28fc  | Mon, 05 Sep 2016 16:36:19 UTC | 1     |
| prod        | website.example.com     | v1    | (git) 6e172a  | Tue, 06 Sep 2016 10:10:00 UTC | 2     |
+-------------+-------------------------+-------+---------------+-------------------------------+-------+
```

We also think it could be interesting for us to have a `project-env-info`, which would display the same as `app-info`:

```
% tranor project-env-info -n website -e dev

(same as app info)
```

fsouza commented:
@tatiana since the build would fail without the dependencies and there's no point in reviewing the vendor directory, I pushed it directly to master and rebased the changes in the PR, it should be easier to review now.

fsouza commented:
@Lepthy down to 11 files now :D

Commit:
Address comments and suggestions in #3

Thanks @aflag, @tatiana and @Lepthy :-D

fsouza commented:
@tatiana should I include the docs in the readme?

Commit:
Add more tests for app related functions

tatiana commented on project.go:
```diff
@@ -0,0 +1,391 @@
+		return fmt.Errorf("failed to configure project %q: %s", c.name, err)
+	}
+	fmt.Fprintf(ctx.Stdout, "successfully created the project %q!\n", c.name)
+	if gitRepo := apps[0]["repository_url"]; gitRepo != "" {
```
Can't we have some issue by using the first repo from the slice `apps`? How are they sorted?

Ideally, we'd have nice-looking git repos URLs for projects, agnostic to environment:
eg:  `git@git.tsuru.example.com:<project-name>.git`

Which would not change in case we wanted to update a project, (ex: add other environments seems a valid use case).

However, it seems this may not be possible, as it may involve changing Tsuru. Could you confirm, @fsouza?

tatiana replied:
Another question: `project-info`  currently doesn't display the `git repo`. Ideally we'd be able to retrieve this information in a project-based repository. What are your thoughts on this?

tatiana commented on project.go:
```diff
@@ -0,0 +1,391 @@
+		c.fs.StringVar(&c.plan, "p", "", "plan to use for the project")
+		c.fs.Var(&c.envs, "envs", "comma-separated list of environments to use (defaults to dev,qa,stage,production)")
+		c.fs.Var(&c.envs, "e", "comma-separated list of environments to use (defaults to env,qa,stage,production)")
+		c.envs.Set("dev,qa,stage,production")
```
Shouldn't we retrieve this from the config file?
Considering we don't retrieve from the config file,  we should isolate this hard-coded environments somewhere else (constant?). This way if anyone wants to use Tranor for other purposes, with a different environment set, they can.
BTW: I strongly suggest we call "production" -> "prod" (everywhere, of course, including the config file - so we keep consistent with an abbreviation for all environments)

tatiana commented on project.go:
```diff
@@ -0,0 +1,391 @@
+	name     string
+	platform string
+	team     string
+	plan     string
```
@fsouza, @aflag, @Lepthy: Do you think it makes sense, within the same project, to have different `plans` per environment?

An example: a certain API in dev & QA may require only 128MB memory, but we may want to be safe in prod and have 512MB.

If you think this use-case makes sense, what do you think would be the best way to set this? The config file would force different apps to have the same `plan` per environment - which doesn't sound good.

aflag replied:
It makes sense for us to have different routers, I think. Each associated with a different ELB, I suppose. And that would be the main reason for me. I think reducing memory for dev environments makes sense as well, but it is more optional.

aflag commented:
We have two issues with the way we are doing pools. One is what @tatiana already flagged. By trying to create a project without passing envs it fails if config.json doesn't have all environments:

```
$ tsuru pool-list 
+-------------------------+--------+-------+
| Pool                    | Kind   | Teams |
+-------------------------+--------+-------+
| dev/dev.example.com     | public |       |
| production/example.com  | public |       |
| qa/qa.example.com       | public |       |
| stage/stage.example.com | public |       |
| theonepool              | public |       |
+-------------------------+--------+-------+
$ tranor env-list
+-------------+------------------+
| Environment | DNS Suffix       |
+-------------+------------------+
| dev         | dev.example.com   |
| stage       | stage.example.com |
| production  | example.com       |
+-------------+------------------+
$ tranor project-create --name myblog --platform python --team admin
Error: failed to load environments: invalid values: qa (valid options are: dev, stage, production)
```

Another problem, which is not directly related to this pull request, is that tsuru is bad at handling pool names with /. Look at my attempt to remove qa/qa.example.com:

```
$ tsuru-admin pool-remove qa/qa.example.com
Are you sure you want to remove "qa/qa.example.com" pool? (y/n) y
Error: 404 page not found
```

tatiana commented:
BTW: will `project-update` be part of this PR as well?

fsouza commented:
@tatiana yes, I'll   include project-update in the PR :)

Commit:
project: use ^c.name in app-list filter

Commit:
Rename production to prod

And also load default environments from config instead of hard coded.

fsouza commented:
@aflag the values for the default enviornments were hard coded, but as suggested by @tatiana, they're now loaded from the config file. Please try again.

aflag commented:
Consider this flow:

```
rafael@zim:~/code/test£ tranor project-create -n melon -l python -t admin 
successfully created the project "melon"!
Git repository: git@192.168.50.4.nip.io:melon-dev.git
rafael@zim:~/code/test£ tsuru app-deploy -a melon-dev .
Uploading files......... ok
/home/application/current /
/

---- Building application image ----
 ---> Sending image to repository (0.02MB)
 ---> Cleaning up

---- Starting 1 new unit [web: 1] ----
 ---> Started unit c7a926ae1d [web]

---- Binding and checking 1 new unit ----
 ---> Bound and checked unit c7a926ae1d [web]

---- Adding routes to new units ----
 ---> Added route to unit c7a926ae1d [web]

---- Setting router healthcheck (Path: /) ----

OK
rafael@zim:~/code/test£ tranor project-list
+------------+--------------+---------------------------+
| Project    | Environments | Address                   |
+------------+--------------+---------------------------+
| melon      | dev          | melon.dev.example.com     |
|            | qa           | melon.qa.example.com      |
|            | stage        | melon.stage.example.com   |
|            | prod         | melon.example.com         |
+------------+--------------+---------------------------+
| strawberry | qa           | strawberry.qa.example.com |
+------------+--------------+---------------------------+
| tomato     | qa           | tomato.qa.example.com     |
|            | prod         | tomato.example.com        |
+------------+--------------+---------------------------+
rafael@zim:~/code/test£ tranor project-info -n melon
Project name: melon

+-------------+-------------------------+-------+-------------------------------+
| Environment | Address                 | Image | Deploy date                   |
+-------------+-------------------------+-------+-------------------------------+
| dev         | melon.dev.example.com   | v1    | Mon, 12 Sep 2016 10:18:25 UTC |
| qa          | melon.qa.example.com    |       |                               |
| stage       | melon.stage.example.com |       |                               |
| prod        | melon.example.com       |       |                               |
+-------------+-------------------------+-------+-------------------------------+
rafael@zim:~/code/test£ tranor project-remove -n melon
Are you sure you want to remove the project "melon"? (y/n) y
rafael@zim:~/code/test£ tranor project-list
+------------+--------------+---------------------------+
| Project    | Environments | Address                   |
+------------+--------------+---------------------------+
| melon      | dev          | melon.dev.example.com     |
+------------+--------------+---------------------------+
| strawberry | qa           | strawberry.qa.example.com |
+------------+--------------+---------------------------+
| tomato     | qa           | tomato.qa.example.com     |
|            | prod         | tomato.example.com        |
+------------+--------------+---------------------------+
```

The reason that melon is still there is that the unit is still being decommissioned. Is it possible for project-remove to only finish after all apps were properly deleted?

aflag commented on project.go:
```diff
@@ -0,0 +1,399 @@
+
+func (c *projectInfo) Info() *cmd.Info {
+	return &cmd.Info{
+		Name: "project-info",
```
Please add the Desc field here.

fsouza replied:
done

aflag commented on project.go:
```diff
@@ -0,0 +1,399 @@
+
+func (c *projectList) Info() *cmd.Info {
+	return &cmd.Info{
+		Name: "project-list",
```
Please add the Desc field here

fsouza replied:
done

aflag commented:
Calling project-info in a project that does not exist should return an error to the user. Instead, this happens:

```
rafael@zim:~/code/test£ tranor project-info -n notfound
Project name: notfound

+-------------+---------+-------+-------------+
| Environment | Address | Image | Deploy date |
+-------------+---------+-------+-------------+
+-------------+---------+-------+-------------+
```

fsouza commented:
@aflag we can write a loop in project-remove that waits for all apps to be removed. Should we do that?

I treated the notfound error on project-remove but forgot to do the same thing on project-info, sorry :-(

Commit:
project: properly report "project not found" on project-info

Commit:
Extract condition into function

Thanks @aflag for the suggestion.

Commit:
Add Desc to project-info and project-list

Commit:
Make project-remove wait untill all apps are removed

Also add some status reporting as apps are deleted.

Commit:
Add description to project-create

Commit:
Improve project-info output

Here's an example of output (extracted from tests):

Project name: proj1
Description: my nice project
Repository: git@example.com:proj1-dev.git
Platform: python
Teams: admin, sysop
Owner: webmaster@example.com
Team owner: admin

+-------------+-------------------------+-------+------------------------------------------------+-------------------------------+-------+
| Environment | Address                 | Image | Git hash/tag                                   | Deploy date                   | Units |
+-------------+-------------------------+-------+------------------------------------------------+-------------------------------+-------+
| dev         | proj1.dev.example.com   | v938  | (git) 40244ff2866eba7e2da6eee8a6fc51464c9f604f | Mon, 05 Sep 2016 01:24:25 UTC | 1     |
| qa          | proj1.qa.example.com    |       |                                                |                               | 2     |
| stage       | proj1.stage.example.com |       |                                                |                               | 2     |
| prod        | proj1.example.com       |       |                                                |                               | 5     |
+-------------+-------------------------+-------+------------------------------------------------+-------------------------------+-------+

fsouza commented:
@aflag I'm still working on project-update, but all the previous comments have been addressed. Thanks for reviewing and testing this!

Regarding the pool name, I tested and we could use either `\` or `|`. I prefer the former, but I don't have a good reason, I just prefer :D

tatiana commented:
@fsouza awesome work, thanks! Let's go with the `\` just for lazy shell programming ;)

Commit:
project: use a backslash in pool name

Commit:
app: remove unecessary if condition

The already handles that for us.

Commit:
project: add command project-env-info

This is like `tsuru app-info`.

Commit:
project: set raw output on project-remove

Commit:
app: add function to update app

Commit:
add basic structure for project-update

Will break the build, sorry about that.

Commit:
Implement project-update

This is finally ready for review :)

fsouza commented:
This is finally ready for review again, including `project-update`!
